File: test/artifact/pldi24/rbtree/rb-bu.hs
-- Adapted from https://github.com/leanprover/lean4/blob/IFL19/tests/bench/rbmap.hs
-- Modified to be strict in the Tree fields
--
-- This is a bottom up implementation following Lorenzen and Leijen's "frame limited reuse" implementation (ICFP'22)
import System.Environment

import Data.Bits
import Data.Int
import Data.Word

data Color = Red | Black

data Tree = Leaf
          | Node !Color !Tree !Int !Int !Tree

data Zipper = Done
            | NodeR !Color !Tree !Int !Int !Zipper
            | NodeL !Color !Zipper !Int !Int !Tree

key :: Tree -> Int
key (Node c l k v r) = k
key Leaf             = 0

is_red t =
  case t of
    Node Red _ _ _ _ -> True
    _ -> False

set_black n =
   case n of
    Node _ l k v r -> Node Black l k v r
    e              -> e

rebuild z t =
   case z of
    NodeR c l k v up -> rebuild up (Node c l k v t)
    NodeL c up k v r -> rebuild up (Node c t k v r)
    Done             -> t
 
balance z l k v r =  
   case z of
    NodeR Black l1 k1 v1 z1 -> rebuild z1 (Node Black l1 k1 v1 (Node Red l k v r))
    NodeL Black z1 k1 v1 r1 -> rebuild z1 (Node Black (Node Red l k v r) k1 v1 r1)
   -- red red violation 
    NodeR Red l1 k1 v1 z1 -> case z1 of
      NodeR _ l2 k2 v2 z2 -> balance z2 (Node Black l2 k2 v2 l1) k1 v1 (Node Black l k v r)
      NodeL _ z2 k2 v2 r2 -> balance z2 (Node Black l1 k1 v1 l) k v (Node Black r k2 v2 r2)
      Done                -> Node Black l1 k1 v1 (Node Red l k v r)    
    NodeL Red z1 k1 v1 r1 -> case z1 of
      NodeR _ l2 k2 v2 z2 -> balance z2 (Node Black l2 k2 v2 l) k v (Node Black r k1 v1 r1)
      NodeL _ z2 k2 v2 r2 -> balance z2 (Node Black l k v r) k1 v1 (Node Black r1 k2 v2 r2)
      Done                -> Node Black (Node Red l k v r) k1 v1 r1
    Done -> Node Black l k v r

find :: Tree -> Int -> Int -> Zipper -> Tree  
find t k v z =
   case t of
    Node c l kx vx r -> if (kx < k) then find r k v (NodeR c l kx vx z)
                          else if (kx > k) then find l k v (NodeL c z kx vx r)
                                           else rebuild z (Node c l k v r)
    Leaf -> balance z Leaf k v Leaf


--------------------------------------------------------------------------------------
-- Benchmarking

access :: Tree -> Int -> Tree
access t k
  = find t k k Done

sumAcc :: Tree -> Int -> Int
sumAcc Leaf acc  = acc
sumAcc (Node _ l x _ r) acc = sumAcc r $! (sumAcc l $! (acc + x))

tsum :: Tree -> Int
tsum t = sumAcc t 0

data Sfc = Sfc !Int32 !Int32 !Int32 !Int32


shiftRightLogical :: Int32 -> Int -> Int32
shiftRightLogical i s = fromIntegral (fromIntegral i `shiftR` s :: Word32)

minHeight (Node _ l _ _ r) = 1 + min (minHeight l) (minHeight r)
minHeight Leaf             = 0

maxHeight (Node _ l _ _ r) = 1 + max (maxHeight l) (maxHeight r)
maxHeight Leaf             = 0

top (Node _ _ x _ _) = x
top Leaf             = 0

sfcStep :: Sfc -> (Int,Sfc)
sfcStep (Sfc x y z cnt)
  = let res = x+y+cnt 
        sfc = Sfc (xor y (shiftRightLogical y 9))
                  (z + (shiftL z 3))
                  ((rotateL z 21) + res)
                  (cnt + 1)
    in seq sfc $ seq res $ (fromIntegral res, sfc)

sfcInit :: Int -> Int -> Sfc
sfcInit seed1 seed2
  = let sfc0 = Sfc 0 (fromIntegral seed1) (fromIntegral seed2) 1
        sfc1 = foldl (\s i -> snd (sfcStep s)) sfc0 [1..12]
    in seq sfc1 sfc1


modE :: Int -> Int -> Int
modE i j
  = let m = i `mod` j
    in if (i < 0 && m < 0) 
        then (if (j < 0) then m - j else m + j)
        else m

test :: Int -> Int -> (Tree -> Int -> Tree) -> (Tree,Int)
test n iter access
  = iloop 0 Leaf (sfcInit 42 43)
  where 
    iloop :: Int -> Tree -> Sfc -> (Tree,Int)
    iloop i t sfc
      = if (i >= iter) 
         then (t, fst (sfcStep sfc))
         else let (t',sfc') = nloop 0 t sfc
              in iloop (i+1) t' sfc'

    nloop i t sfc
      = if (i >= n) 
          then (t,sfc)
          else let (x,sfc') = sfcStep sfc
                   t' = access t $! (modE x n)
               in seq (key t') $ seq sfc' $
                  nloop (i+1) t' sfc'

benchMain :: (Tree -> Int -> Tree) -> IO ()
benchMain access
  = do args <- getArgs 
       let n = case args of
                 arg:_ -> read arg
                 _     -> 100000
       let sfc0 = sfcInit 42 43
       -- putStrLn ("init: " ++ show (fst (sfcStep sfc0)))
       let (t,final) = test n 100 access
       putStrLn ("sum: " ++ show (tsum t) ++ ", height: " ++ show (maxHeight t) ++ "/" ++ show (minHeight t) ++ ", top: " ++ show (top t) ++ ", final acces: " ++ show final)


main :: IO ()
main 
  = benchMain access

File: test/artifact/pldi24/rbtree/rb-bu.ml
(* Adapted from https://github.com/leanprover/lean4/blob/IFL19/tests/bench/rbmap.ml *)
(* This is a bottom up implementation following Lorenzen and Leijen's "frame limited reuse" implementation (ICFP'22) *)

open Int32;;

type color =
| Red
| Black;;

type tree =
| Leaf
| Node of color * tree * int * int * tree;;

type zipper =
| Done
| NodeR of color * tree * int * int * zipper
| NodeL of color * zipper * int * int * tree;;


let is_red t =
  match t with
  | Node (Red, _, _, _, _) -> true
  | _ -> false;;

let set_black n =
  match n with
  | Node (_, l, k, v, r) -> Node (Black, l, k, v, r)
  | e                    -> e;;

let rec rebuild z t =
  match z with
  | NodeR(c,l,k,v,up) -> rebuild up (Node(c,l,k,v,t))
  | NodeL(c,up,k,v,r) -> rebuild up (Node(c,t,k,v,r))
  | Done              -> t;;
 
let rec balance z l k v r =  
  match z with
  | NodeR(Black,l1,k1,v1,z1) -> rebuild z1 (Node(Black,l1,k1,v1,Node(Red,l,k,v,r)))
  | NodeL(Black,z1,k1,v1,r1) -> rebuild z1 (Node(Black,Node(Red,l,k,v,r),k1,v1,r1))
  (* red red violation *)
  | NodeR(Red,l1,k1,v1,z1) -> begin match z1 with
    | NodeR(_,l2,k2,v2,z2) -> balance z2 (Node(Black,l2,k2,v2,l1)) k1 v1 (Node(Black,l,k,v,r))
    | NodeL(_,z2,k2,v2,r2) -> balance z2 (Node(Black,l1,k1,v1,l)) k v (Node(Black,r,k2,v2,r2))
    | Done                 -> Node(Black,l1,k1,v1,Node(Red,l,k,v,r))
    end
  | NodeL(Red,z1,k1,v1,r1) -> begin match z1 with
    | NodeR(_,l2,k2,v2,z2) -> balance z2 (Node(Black,l2,k2,v2,l)) k v (Node(Black,r,k1,v1,r1))
    | NodeL(_,z2,k2,v2,r2) -> balance z2 (Node(Black,l,k,v,r)) k1 v1 (Node(Black,r1,k2,v2,r2))
    | Done                 -> Node(Black,Node(Red,l,k,v,r),k1,v1,r1)
    end
  | Done -> Node(Black,l,k,v,r);;
  
let rec find t k v z =
  match t with
  | Node(c,l,kx,vx,r) -> if (kx < k) then find r k v (NodeR(c,l,kx,vx,z))
                          else if (kx > k) then find l k v (NodeL(c,z,kx,vx,r))
                                           else rebuild z (Node(c,l,k,v,r))
  | Leaf -> balance z Leaf k v Leaf;;

let access t k =
  find t k k Done
  (* if is_red t then set_black (find t k k Done)
     else ins t k k;; *)

let rec fold f n d =
match n with
| Leaf -> d
| Node(_, l, k, v, r) -> fold f r (f k v (fold f l d));;


(* helpers *)

let rec sumacc t acc = 
  match t with
  | Node (c,l,k,x,r) -> sumacc r (sumacc l (acc+x))
  | Leaf           -> acc;;

let sum t = sumacc t 0;;


let imin x y = if (x <= y) then x else y;;
let imax x y = if (x >= y) then x else y;;

let rec minheight t = 
  match t with
  | Node (_,l,_,_,r) -> 1 + imin (minheight l) (minheight r)
  | Leaf             -> 0;;

let rec maxheight t = 
  match t with
  | Node (_,l,_,_,r) -> 1 + imax (maxheight l) (maxheight r)
  | Leaf             -> 0;;

let top t = 
  match t with
  | Node (_,l,x,_,r) -> x
  | Leaf             -> 0;;


type sfc =
| Sfc of int32 * int32 * int32 * int32;;

let rotl i n =
  logor (shift_left i n) (shift_right_logical i (32 - n));;

let sfc_step sfc =
  match sfc with
  | Sfc (x,y,z,cnt) -> 
    let res = add x  (add y cnt) in
    (to_int res,Sfc (logxor y (shift_right_logical y 9),
                      add z (shift_left z 3),
                      add (rotl z 21)  res,
                      add cnt one ));;

let sfc_init seed1 seed2 =
  let s = ref (Sfc (zero,of_int seed1, of_int seed2,one)) in
  for i = 1 to 12 do
    let (_,s1) = sfc_step !s in
    s := s1
  done;
  !s;;


let modE i j =
  let m = i mod j in
  if (i < 0 && m < 0) then (if (j < 0) then m - j else m + j)
  else m;;

let test n iter =
  let t = ref Leaf in
  let s = ref (sfc_init 42 43) in
  for i = 1 to iter do
    for j = 1 to n do
      let (x,s1) = sfc_step !s in
      s := s1;
      t := access !t (modE x n) 
    done;
  done;
  let (final,_) = sfc_step !s in
  (!t,final);;

let main n = 
  let (t,final) = test (if n == 0 then 100000 else n) 100 in
  Printf.printf "sum: %d, height: %d/%d, top: %d, final access: %d\n" (sum t) (maxheight t) (minheight t) (top t) final;;

main (int_of_string Sys.argv.(1));; 
